test/e2e_node/container_lifecycle_test.go

```diff
@@ -2415,6 +2415,39 @@ var _ = SIGDescribe(nodefeature.SidecarContainers, "Containers Lifecycle", func(
 			ginkgo.It("should start a regular container", func() {
 				framework.ExpectNoError(results.HasRestarted(regular1))
 			})
+
+			ginkgo.It("should not restart when updated with a new image", func() {
+				client := e2epod.NewPodClient(f)
+				podSpec, err := client.Get(context.TODO(), podSpec.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+				results = parseOutput(context.TODO(), f, podSpec)
+
+				tStartRestartableInit1, err := results.TimeOfStart(restartableInit1)
+				framework.ExpectNoError(err)
+				tStartRegular1, err := results.TimeOfStart(regular1)
+				framework.ExpectNoError(err)
+
+				client.Update(context.TODO(), podSpec.Name, func(pod *v1.Pod) {
+					pod.Spec.InitContainers[0].Image = "fakeimage"
+				})
+				framework.ExpectNoError(err)
+
+				updatedPodSpec, err := client.Get(context.TODO(), podSpec.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+				results = parseOutput(context.TODO(), f, updatedPodSpec)
+
+				tStartUpdatedRestartableInit1, err := results.TimeOfStart(restartableInit1)
+				framework.ExpectNoError(err)
+				tStartUpdatedRegular1, err := results.TimeOfStart(regular1)
+				framework.ExpectNoError(err)
+
+				ginkgo.By("ensuring not restarted the regular container", func() {
+					gomega.Expect(tStartUpdatedRegular1).To(gomega.Equal(tStartRegular1))
+				})
+				ginkgo.By("ensuring not restarted the init container", func() {
+					gomega.Expect(tStartUpdatedRestartableInit1).To(gomega.Equal(tStartRestartableInit1))
```

Comment:
this is interesting. So if we change image to something bad, it will not be restarted. What will happen? Will the container run?

For the regular container - if image is changed, the new image will be used, so there will be a restart (not sure if restart count will increment). If the bad image is assigned, the container will start failing.

So I am not sure why this test passes

Reply:
> So if we change image to something bad, it will not be restarted. What will happen? Will the container run?

Only the pod spec is updated when the init container's image is updated to invalid one, and all containers in the pod continue to run.

```console
❯ kind create cluster
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.30.0) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Have a nice day! 👋

❯ cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  initContainers:
  - name: init1
    image: busybox:1.36.0
    command: ["sleep", "infinity"]
    restartPolicy: Always
  containers:
  - name: regular1
    image: busybox:1.36.0
    command: ["sleep", "infinity"]
  - name: regular2
    image: nginx:1.14.1
    ports:
    - containerPort: 80
EOF
pod/nginx created

❯ kubectl wait pods nginx --for=jsonpath='{.status.phase}'=Running
pod/nginx condition met

# Update to non-existent image
❯ kubectl set image pod/nginx init1=busybox:1.36.100
pod/nginx image updated

❯ kubectl get pod nginx -o jsonpath='{.spec.initContainers[?(@.name=="init1")].image}'
busybox:1.36.100

❯ kubectl get pod nginx -o jsonpath='{.status.initContainerStatuses[?(@.name=="init1")].image}'
docker.io/library/busybox:1.36.0

❯ kubectl describe pods nginx
(...)
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  85s   default-scheduler  Successfully assigned default/nginx to kind-control-plane
  Normal  Pulled     84s   kubelet            Container image "busybox:1.36.0" already present on machine
  Normal  Created    84s   kubelet            Created container init1
  Normal  Started    84s   kubelet            Started container init1
  Normal  Pulled     83s   kubelet            Container image "busybox:1.36.0" already present on machine
  Normal  Created    83s   kubelet            Created container regular1
  Normal  Started    83s   kubelet            Started container regular1
  Normal  Pulled     83s   kubelet            Container image "nginx:1.14.1" already present on machine
  Normal  Created    83s   kubelet            Created container regular2
  Normal  Started    83s   kubelet            Started container regular2

❯ kubectl get pods
NAME    READY   STATUS    RESTARTS   AGE
nginx   3/3     Running   0          5m46s
```

> so there will be a restart (not sure if restart count will increment)

The restart count of the regular container is incremented when its image is updated to valid one.

```console
❯ cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  initContainers:
  - name: init1
    image: busybox:1.36.0
    command: ["sleep", "infinity"]
    restartPolicy: Always
  containers:
  - name: regular1
    image: busybox:1.36.0
    command: ["sleep", "infinity"]
  - name: regular2
    image: nginx:1.14.1
    ports:
    - containerPort: 80
EOF
pod/nginx created

❯ kubectl wait pods nginx --for=jsonpath='{.status.phase}'=Running
pod/nginx condition met

❯ kubectl set image pod/nginx regular1=busybox:1.36.1
pod/nginx image updated

❯ kubectl get pod nginx -o jsonpath='{.status.containerStatuses[?(@.name=="regular1")].image}'
docker.io/library/busybox:1.36.1

❯ kubectl get pods
NAME    READY   STATUS    RESTARTS     AGE
nginx   3/3     Running   1 (2s ago)   44s
```

However, the restart won't happen and will start a crash loop backoff (image pull backoff) if its image is updated to invalid one.

```console
❯ cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  initContainers:
  - name: init1
    image: busybox:1.36.0
    command: ["sleep", "infinity"]
    restartPolicy: Always
  containers:
  - name: regular1
    image: busybox:1.36.0
    command: ["sleep", "infinity"]
  - name: regular2
    image: nginx:1.14.1
    ports:
    - containerPort: 80
EOF
pod/nginx created

❯ kubectl wait pods nginx --for=jsonpath='{.status.phase}'=Running
pod/nginx condition met

# Update to non-existent image
❯ kubectl set image pod/nginx regular1=busybox:1.36.100
pod/nginx image updated

❯ kubectl get pod nginx -o jsonpath='{.status.containerStatuses[?(@.name=="regular1")].image}'
docker.io/library/busybox:1.36.0

❯ kubectl describe pod nginx
(...)
  Normal   Killing    39s   kubelet            Container regular1 definition changed, will be restarted
  Normal   Pulling    9s    kubelet            Pulling image "busybox:1.36.100"
  Warning  Failed     7s    kubelet            Failed to pull image "busybox:1.36.100": rpc error: code = NotFound desc = failed to pull and unpack image "docker.io/library/busybox:1.36.100": failed to resolve reference "docker.io/library/busybox:1.36.100": docker.io/library/busybox:1.36.100: not found
  Warning  Failed     7s    kubelet            Error: ErrImagePull
  Normal   BackOff    6s    kubelet            Back-off pulling image "busybox:1.36.100"
  Warning  Failed     6s    kubelet            Error: ImagePullBackOff

# Start image pull backoff and crash loop backoff alternately
❯ kubectl get pods
NAME    READY   STATUS             RESTARTS   AGE
nginx   2/3     ImagePullBackOff   0          94s

❯ kubectl get pods
NAME    READY   STATUS             RESTARTS   AGE
nginx   2/3     CrashLoopBackOff   0          2m
```

Reply:
I would think this is a bug. Allowing customers to update sidecar's image would be useful. I thought @tzneal manually verified it worked. Not sure if this broke somehow